### PR TITLE
feat: Auto-Populate POSIX Information on sign in using SSO [CM-399]

### PR DESCRIPTION
**Pull request by kkunapuli**
## Ticket

[CM-399]

## Description

When the a determined master is configured to expected the POSIX claims in SSO login & they are provided in the claim, then we will populate the relevant fields for the user:

	AgentUID   null.Int    `db:"agent_uid" json:"agent_uid"`
	AgentGID   null.Int    `db:"agent_gid" json:"agent_gid"`
	AgentUser  null.String `db:"agent_user" json:"agent_user"`
	AgentGroup null.String `db:"agent_group" json:"agent_group"`

Also re-adds new fields to master configuration (original done here https://github.com/determined-ai/determined/pull/9690/files). The original PR was reverted because of CI errors. 

## Test Plan

Test using an Okta application.

### OIDC
Recommended application: `carolina-oidc-test`, https://dev-2564556-admin.okta.com/admin/app/oidc_client/instance/0oacxxzecx7a9l5O85d7/#tab-general 

- add your user to the test application, filling in the fields `posix_uid` (int), `posix_gid` (int), `agent_user_name` (string) and `agent_group_name` (string)
- in your master config, add the following fields to an `oidc` section:
```
oidc:
  agent_uid_attribute_name: posix_uid
  agent_gid_attribute_name: posix_gid
  agent_user_name_attribute_name: agent_user_name
  agent_group_name_attribute_name: agent_group_name
  always_redirect: true
```
- start devcluster, and log in using Okta
- inspect the database and ensure that it is updated, e.g. `SELECT * FROM agent_user_groups;` and look at columns `uid`, `gid`, `user_` and `group_`

**You can also try different combinations! A user should be able to specify any, none, or all of the new parameters.**

### SAML
Recommended application: `Determined SAML CI`, https://dev-2564556-admin.okta.com/admin/app/dev-2564556_determinedsamlci_1/instance/0oaganmj6zxHC3A575d7/#tab-assignments

- add your user to the test application, the fields will be filled out automatically
- in your master config, add the following fields to an `saml` section:
```
saml:
  agent_uid_attribute_name: posix_uid
  agent_gid_attribute_name: posix_gid
  agent_user_name_attribute_name: agent_user_name
  agent_group_name_attribute_name: agent_group_name
  always_redirect: true
```
- start devcluster, and log in using Okta
- inspect the database and ensure that it is updated, e.g.
```
determined=# select * from agent_user_groups;
 id | user_id |  user_   | uid |  group_  | gid 
----+---------+----------+-----+----------+-----
  2 |       3 | Kristine |   4 | Kunapuli |   5
```

`user_` will be your first name and `group_` will be your last name.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] New features have been approved by the corresponding PM
- [ ] User-facing API changes have the "User-facing API Change" label
- [ ] Release notes have been added as a separate file under `docs/release-notes/`
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses have been included for new code which was copied and/or modified from any external code

[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[CM-399]: https://hpe-aiatscale.atlassian.net/browse/CM-399?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: master/internal/config/saml_config.go [hunk 1]
```diff
@@ -17,8 +17,8 @@ type SAMLConfig struct {
 	AutoProvisionUsers          bool   `json:"auto_provision_users"`
 	GroupsAttributeName         string `json:"groups_attribute_name"`
 	DisplayNameAttributeName    string `json:"display_name_attribute_name"`
-	AgentUIDAttributeName       int    `json:"agent_uid_attribute_name"`
-	AgentGIDAttributeName       int    `json:"agent_gid_attribute_name"`
+	AgentUIDAttributeName       string `json:"agent_uid_attribute_name"`
+	AgentGIDAttributeName       string `json:"agent_gid_attribute_name"`
```
Comment (kkunapuli) [inline master/internal/config/saml_config.go:21]: Being an "attribute name" I think string makes more sense here.

File: master/internal/plugin/saml/service.go [hunk 2]
```diff
@@ -227,10 +236,25 @@ func (s *Service) toUserAttributes(response *saml.Assertion) *userAttributes {
 		return nil
 	}
 
+	strAgentUID := getSAMLAttribute(response, string(s.userConfig.agentUIDAttributeName))
+	tempAgentUID, err := strconv.Atoi(strAgentUID)
+	if err != nil {
+		logrus.WithError(err).WithField("agentUID", strAgentUID).Error("unable to convert to integer")
```
Comment (kkunapuli) [inline master/internal/plugin/saml/service.go:242]: I think this error should be very rare / surprising but we should log it just in case.

File: master/internal/config/saml_config.go (hunk 1 above)
Comment (corban-beaird) [reply master/internal/config/saml_config.go:21]: That's a fair point! I wonder why it was this way to begin with?

Comment (corban-beaird) [reply master/internal/config/saml_config.go:21]: Oh it was because it wasn't originally the attribute names? It was the values, and we later shifted that

File: master/internal/user/postgres_agentusergroup.go [hunk 3]
```diff
@@ -56,7 +57,9 @@ func GetAgentUserGroup(
 	workspaceID int,
 ) (*model.AgentUserGroup, error) {
 	workspaceAug, err := getAgentUserGroupFromWorkspaceID(ctx, workspaceID)
-	if err != nil {
+	if err == sql.ErrNoRows {
+		logrus.WithError(err).Warnf("no agent user group results from workspaceID=%d", workspaceID)
+	} else if err != nil {
```
Comment (kkunapuli) [inline master/internal/user/postgres_agentusergroup.go:62]: `workspaceID` is supposed to be optional, but bun throws an error if zero results are returned.

File: docs/release-notes/auto-populate-posix.rst [hunk 4]
```diff
@@ -0,0 +1,9 @@
+:orphan:
+
+**New Features**
+
+-  Master Configuration: Master Configuration accepts POSIX claims `agent_uid_attribute_name`,
+   `agent_gid_attribute_name`, `agent_user_name_attribute_name`, or
+   `agent_group_name_attribute_name`. Please visit :ref:`oidc master configuration
+   <_master-config-oidc>` or :ref:`saml master configuration <_master-config-saml>` for details. If
+   any of the fields are configured, they will be used and synced to the database.
```
Comment (kkunapuli) [inline docs/release-notes/auto-populate-posix.rst:9]: I don't understand why it's accepting `_master-config-saml` as a reference link but not `_master-config-oidc`.

File: master/internal/config/saml_config.go (hunk 1 above)
Comment (kkunapuli) [reply master/internal/config/saml_config.go:21]: Yes, exactly!

File: docs/release-notes/auto-populate-posix.rst (hunk 4 above)
Comment (tara-det-ai) [reply docs/release-notes/auto-populate-posix.rst:9]: :orphan:

**New Features**

-  Master Configuration: Add support for POSIX claims in the master configuration. It now accepts 
   ``agent_uid_attribute_name``, ``agent_gid_attribute_name``, ``agent_user_name_attribute_name``, 
   or ``agent_group_name_attribute_name``. Refer to the :ref:`OIDC master configuration <master-config-oidc>` 
   or :ref:`SAML master configuration <master-config-saml>` for details. If any of these fields 
   are configured, they will be used and synced to the database.

Comment (tara-det-ai) [reply docs/release-notes/auto-populate-posix.rst:9]: `:orphan:

**New Features**

-  Master Configuration: Add support for POSIX claims in the master configuration. It now accepts 
   ``agent_uid_attribute_name``, ``agent_gid_attribute_name``, ``agent_user_name_attribute_name``, 
   or ``agent_group_name_attribute_name``. Refer to the :ref:`OIDC master configuration <master-config-oidc>` 
   or :ref:`SAML master configuration <master-config-saml>` for details. If any of these fields 
   are configured, they will be used and synced to the database.
`

Comment (tara-det-ai) [reply docs/release-notes/auto-populate-posix.rst:9]: @kkunapuli i think maybe you added the leading underscore and its that part that is not needed. maybe something with the angle brackets. https://docs.readthedocs.io/en/stable/guides/cross-referencing-with-sphinx.html

**Review by tara-det-ai — COMMENTED** [review]
docs stamp not required to merge; left a suggested edit

File: master/internal/plugin/saml/service.go (hunk 2 above)
Comment (hamidzr) [reply master/internal/plugin/saml/service.go:238]: wdy think about having this parser return an err? the only place it's used does support returning an error that might be the right behavior here.

**Review by hamidzr — APPROVED** [review]

File: docs/release-notes/auto-populate-posix.rst (hunk 4 above)
Comment (kkunapuli) [reply docs/release-notes/auto-populate-posix.rst]: That was it - thank you so much @tara-det-ai!

File: master/internal/plugin/oidc/service.go [hunk 5]
```diff
@@ -272,19 +309,52 @@ func (s *Service) lookupUser(ctx context.Context, claimValue string) (*model.Use
...
+	}
+	if sessionData.AgentGID != 0 {
+		result.GID = sessionData.AgentGID
+	}
+	if sessionData.AgentUserName != "" {
+		result.User = sessionData.AgentUserName
+	}
+	if sessionData.AgentGroupName != "" {
+		result.Group = sessionData.AgentGroupName
+	}
```
Comment (varlogtim) [inline master/internal/plugin/oidc/service.go:331]: I am guessing that we are checking the `AgentUID` and `AgentGID` are not equal to `0` because golang auto initializes ints that aren't assigned values to `0` when in reality we want to check to see if any value, including `0`, has been set in the `IDTokenClaims.AgentUID` and `IDTokenClaims.AgentGID`.

While this likely won't happen, I imagine a situation in which a user had a specific non-zero UID and GID set for their user. The user is then promoted to a "sys admin" or something and they now need to run tasks as `root`. If I am reading this code correctly, in this situation we wouldn't update the UID and GID of the user when the auth claim returned contained a AgentUID and AgentGID of `0`.

If I am correct, I think we might need to do something like use a int pointer and check if `nil` vs. checking if `0` to handle situations in which the admin wants to intentionally give someone permission to run as root.

**Review by varlogtim — COMMENTED** [review]
Possible concern.

File: master/internal/plugin/oidc/service.go [hunk 6]
```diff
@@ -272,19 +309,52 @@ func (s *Service) lookupUser(ctx context.Context, claimValue string) (*model.Use
...
 
+func mergeUserGroups(sessionData *IDTokenClaims, dbData *model.AgentUserGroup) *model.AgentUserGroup {
+	result := model.AgentUserGroup{
+		UID:   dbData.UID,
+		GID:   dbData.GID,
+		User:  dbData.User,
+		Group: dbData.Group,
+	}
+
+	if sessionData.AgentUID != 0 {
```
Comment (varlogtim) [inline master/internal/plugin/oidc/service.go:320]: If I am reading this correctly, it looks like we initialize all the `result` with the UID and GID that already exist for the user in the DB.

I am guessing that we are checking the AgentUID and AgentGID are not equal to 0 because golang auto initializes ints that aren't assigned values to 0 when in reality we want to check to see if any value, including 0, has been set in the IDTokenClaims.AgentUID and IDTokenClaims.AgentGID.

While this likely won't happen, I imagine a situation in which a user had a specific non-zero UID and GID set for their user. The user is then promoted to a "sys admin" or something and they now need to run tasks as root. If I am reading this code correctly, in this situation we wouldn't update the UID and GID of the user when the auth claim returned contained a AgentUID and AgentGID of 0.

If I am correct, I think we might need to do something like use a int pointer and check if nil vs. checking if 0 to handle situations in which the admin wants to intentionally give someone permission to run as root.

**Review by varlogtim — COMMENTED** [review]
Possible issue

File: master/internal/plugin/oidc/service.go (hunk 6 above)
Comment (kkunapuli) [reply master/internal/plugin/oidc/service.go:320]: Yes, the default value for an int in Golang is 0. You're absolutely right that, as is, it is not possible to reset the agent uid/gid to 0 - great catch! 

I don't know about the plausibility of the specific use case mentioned. I'm going to start a conversation in a team channel so people with more experience with this part of the code can weigh in.

File: master/internal/plugin/saml/service.go (hunk 2 above)
Comment (kkunapuli) [reply master/internal/plugin/saml/service.go]: Great idea, thanks!

File: master/internal/plugin/oidc/service.go (hunk 6 above)
Comment (varlogtim) [reply master/internal/plugin/oidc/service.go]: Cool, I see a couple of bool vars were added to handle this. Thanks!

File: master/internal/plugin/oidc/service.go [hunk 7]
```diff
@@ -272,19 +313,52 @@ func (s *Service) lookupUser(ctx context.Context, claimValue string) (*model.Use
...
 
+func mergeUserGroups(sessionData *IDTokenClaims, dbData *model.AgentUserGroup) *model.AgentUserGroup {
+	result := model.AgentUserGroup{
+		UID:   dbData.UID,
+		GID:   dbData.GID,
+		User:  dbData.User,
+		Group: dbData.Group,
+	}
+
+	if sessionData.AgentUIDSet {
```
Comment (varlogtim) [inline master/internal/plugin/oidc/service.go:324]: Another nit with this pattern.

I am imagining the following UX:

TL;DR: once a user as a UID/GID set, it can never be unset via Auth claims and would need to be manually removed.

1. System admin configures `security.default_task.user_id: 1234`, in `master.yaml 
2. Some impetus: results in `UserA` having AgentUID `5432` being set in the auth claim.
3. Other impetus: results in `UserA`'s AgentUID being removed from the auth claim _for which the admin would expect that tasks run by that user would run as UID 1234, however, I suspect they will continue to run as 5432._

To be clear, I am OK with this behavior, I think... but I wanted to bring it up here for the following reasons:
1. To get confirmation that I speculation about the behavior here is correct.
2. Document that this is the behavior we should expect.
3. Get any input from folks other than me if this behavior is acceptable.

Comment (varlogtim) [reply master/internal/plugin/oidc/service.go:324]: I tested this behavior today and confirmed that it works as I expected in might.

_Once you have enabled this feature and an AgentUID/GID is set for a user in order to have have this revert back to using the `security.default_task.uid` an admin needs to manually remove the AgentUID/GID settings for that user._

We discussed real would situations in which this type of thing could come and yesterday and the only real one we could think is if a customer switches IdPs and forgets to populate the Agent UID/GID attributes in the auth claim. In this situation we would likely want to keep the UID and GID of the users there and not remove them and have the system go back to user the default. Therefore, I think the current behavior is best.

File: docs/manage/users-remote.rst [hunk 8]
```diff
@@ -77,10 +81,24 @@ enable user auto-provisioning and the remote management of any information attac
...
+Custom claims can be recognized and automatically linked by configuring the attribute name with the
+relevant field, e.g. ``agent_uid_attribute_name: "custom_uid_attribute_name"``.
+
+.. note::
+
+   Once this feature is enabled and an agent UID/GID is set for a user, the UID/GID can only be
+   reset manually. In order to have have the UID or GID revert back to using the
+   security.default_task.uid(gid), an admin needs to manually remove the agent UID/GID settings for
+   that user.
+
```
Comment (tara-det-ai) [inline docs/manage/users-remote.rst:101]: To automatically link custom claims, you can configure the attribute name with the relevant field. For example:

``agent_uid_attribute_name: "custom_uid_attribute_name"``

.. note::

   Once this feature is enabled and an agent UID/GID is assigned to a user, it can only be reset manually. 
   To revert the UID or GID back to the default value specified by ``security.default_task.uid(gid)``, an admin 
   must manually remove the agent UID/GID settings for that user.

**Review by tara-det-ai — COMMENTED** [review]
suggested edit
